src/Factory/Api/Mastodon/Attachment.php

```diff
@@ -51,39 +51,63 @@ public function createFromUriId(int $uriId): array
 	{
 		$attachments = [];
 		foreach (Post\Media::getByURIId($uriId, [Post\Media::AUDIO, Post\Media::VIDEO, Post\Media::IMAGE]) as $attachment) {
-			$filetype = !empty($attachment['mimetype']) ? strtolower(substr($attachment['mimetype'], 0, strpos($attachment['mimetype'], '/'))) : '';
-
-			if (($filetype == 'audio') || ($attachment['type'] == Post\Media::AUDIO)) {
-				$type = 'audio';
-			} elseif (($filetype == 'video') || ($attachment['type'] == Post\Media::VIDEO)) {
-				$type = 'video';
-			} elseif ($attachment['mimetype'] == 'image/gif') {
-				$type = 'gifv';
-			} elseif (($filetype == 'image') || ($attachment['type'] == Post\Media::IMAGE)) {
-				$type = 'image';
-			} else {
-				$type = 'unknown';
-			}
+			$attachments[] = $this->createFromMediaArray($attachment);
+		}
+
+		return $attachments;
+	}
+
+	/**
+	 * @param int $id id of the media
+	 * @return array
+	 * @throws HTTPException\InternalServerErrorException
+	 */
+	public function createFromId(int $id): array
```

Comment:
```suggestion
	public function createFromId(int $id): \Friendica\Object\Api\Mastodon\Attachment
```

Reply:
Please have a look at the function `createFromUriId` as well. Here we assign the content from `createFromMediaArray` as well.

Reply:
It's fine, it's populating a list with objects now instead of arrays.

Reply:
I will test it at the weekend. But nonetheless, at least for me it doesn't improve the readability of the code. In fact it adds some "WTF is happening here?" for me.

Reply:
Even if removes some steps to convert it to array?

Let's go back to the spirit of data transfer objects and their factories: the goal of data transfer objects is to have a rigid data structure, not arbitrary arrays that can be changed at any point in the code without notice. It's especially important for APIs for which we do not decide the return structure like the Mastodon one.

The goal of the factories to make data transfer object instantiation easier than always having to call a constructor with ten parameters or so. Factories provide context-relevant methods to instantiate data transfer objects based on the data we have at hand. These methods will return instances of their corresponding data transfer object or lists of them.

Once you have created a data transfer object using its factory, you can pass it around to its final destination, but, and it's its entire purpose, you cannot modify it, ensuring the integrity of the resulting API call that makes use of it.

Reply:
Well, in fact we have to remove fields in the API depending upon the data, see the `toArray` calls in these classes.

I've got a problem with this "magic" that is executed automagically during these whole conversion, since in case of problems I can't really trace it.

Reply:
The business logic in the `toArray` methods isn't ideal, but at least it's only happening in one place, so it will be consistent no matter the API call.

Reply:
I still consider this as a way to make the code less readable, traceable and understandable (at least for me). I applied the changes and I get no obvious errors, but I really don't like it.

Reply:
Thank for following along. It does leverage some language magic, which I personally like because it reduces the actual code written, but I understand it isn't the same for everybody.